app/gui/src/dashboard/providers/ModalProvider.tsx

```diff
@@ -25,14 +33,94 @@ interface ModalContextType {
   readonly modal: Modal | null
 }
 
+const ModalsStore = createStore<{
+  readonly key: number
+  readonly modal: Modal | null
+  readonly setModal: (modal: ModalOrCallback | null) => void
+  readonly updateModal: (modal: ModalOrCallback) => void
+}>((set, get) => ({
+  key: 0,
+  modal: null,
+  setModal: (modal) => {
+    const existingModal = get().modal
+
+    const nextKey = get().key + 1
+
+    if (typeof modal === 'function') {
+      set({ modal: modal(existingModal), key: nextKey })
+    } else {
+      set({ modal, key: nextKey })
+    }
+  },
+  updateModal: (modal) => {
+    const existingModal = get().modal
+
+    if (existingModal == null) {
+      throw new Error('Calling updateModal while no modal is set is forbidden.')
+    }
+
+    if (typeof modal === 'function') {
+      set({ modal: modal(existingModal) })
+    } else {
+      set({ modal })
+    }
+  },
+}))
+
+/**
+ * Set the currently active modal.
+ * @throws An error if a modal is already set.
+ */
+export function setModal(modal: ModalOrCallback) {
+  const modalsStore = ModalsStore.getState()
+  modalsStore.setModal(modal)
+}
+
+/**
+ * Update the currently active modal.
+ * @throws An error if no modal is set.
+ */
+export function updateModal(modal: ModalOrCallback) {
+  const modalsStore = ModalsStore.getState()
+  modalsStore.updateModal(modal)
+}
+
+/**
+ * Unset the currently active modal.
+ */
+export function unsetModal() {
+  const modalsStore = ModalsStore.getState()
+  modalsStore.setModal(null)
+}
+
+/**
+ * Get the currently active modal.
+ */
+export function getModal() {
+  const modalsStore = ModalsStore.getState()
+  return modalsStore.modal
+}
+
 const ModalContext = React.createContext<ModalContextType>({ modal: null, key: 0 })
 
 const ModalStaticContext = React.createContext<ModalStaticContextType>({
-  setModal: () => {
-    // Ignored. This default value will never be used as `ModalProvider` always provides
-    // its own value.
-  },
-  modalRef: { current: null },
+  setModal: ModalsStore.getState().setModal,
+  modalRef: new Proxy(
+    { current: ModalsStore.getState().modal },
+    {
+      /**
+       * Makes the `modalRef` always return the latest modal.
```

Comment:
any reason we aren't just doing `{ get current() { return ModalsStore.getState().modal } }` instead of a proxy here?

Reply:
Yeah I think getters are better here.